Add an option for verbose stats at the end of the match

KingOfSquares opened the pull request:
This ended up being bigger then I initially thought..

i refactored how the InventoryMenu works a little in this as well.

~~I'm not completely done with testing(there is a bug with stacking listeners on pagination if I remember correctly) but I thought of putting this up to get some roasts on my code.~~
Edit: this seems to have been fixed

~~I will add pictures before this weekend has finished~~

An example: https://streamable.com/310629

applenick commented on core/src/main/java/tc/oc/pgm/menu/InventoryMenuUtils.java:
```diff
@@ -0,0 +1,160 @@
+   *     this menu
+   * @return a pretty menu that fits any amount of {@link InventoryMenuItem}s
+   */
+  public static InventoryMenu prettyMenu1(
```
Maybe instead of prettyMenu1 2 3 etc, you could make it more informative. For example

1-> `prettyMenu()`

2-> `smallMenu()`

3 -> `normalMenu()`

Feel free to use other names, but just so you get the general idea 😄

applenick commented on core/src/main/java/tc/oc/pgm/observers/tools/FlySpeedTool.java:
```diff
@@ -17,6 +17,12 @@
 
   private static String TRANSLATION_KEY = "setting.flyspeed.";
 
+  InventoryMenu in;
+
+  public FlySpeedTool(InventoryMenu in) {
+    this.in = in;
```
```suggestion
  InventoryMenu menu;

  public FlySpeedTool(InventoryMenu menu) {
    this.menu = menu;
```

KingOfSquares replied:
Wops this is just code I forgot to remove

applenick commented on core/src/main/java/tc/oc/pgm/observers/tools/GamemodeTool.java:
```diff
@@ -20,6 +20,12 @@
 
 public class GamemodeTool implements InventoryMenuItem {
 
+  final InventoryMenu in;
+
+  public GamemodeTool(InventoryMenu in) {
+    this.in = in;
```
```suggestion
  final InventoryMenu menu;

  public GamemodeTool(InventoryMenu menu) {
    this.menu = menu;
```

KingOfSquares replied:
Wops this is just code I forgot to remove

applenick commented on core/src/main/java/tc/oc/pgm/observers/tools/NightVisionTool.java:
```diff
@@ -17,6 +17,12 @@
 
 public class NightVisionTool implements InventoryMenuItem {
 
+  final InventoryMenu in;
+
+  public NightVisionTool(InventoryMenu in) {
+    this.in = in;
```
```suggestion
  final InventoryMenu menu;

  public NightVisionTool(InventoryMenu menu) {
    this.menu = menu;
```

KingOfSquares replied:
Wops this is just code I forgot to remove

applenick commented on core/src/main/java/tc/oc/pgm/observers/tools/VisibilityTool.java:
```diff
@@ -18,6 +18,12 @@
 
 public class VisibilityTool implements InventoryMenuItem {
 
+  final InventoryMenu in;
+
+  public VisibilityTool(InventoryMenu in) {
+    this.in = in;
```
```suggestion
  final InventoryMenu menu;

  public VisibilityTool(InventoryMenu menu) {
    this.menu = menu;
```

KingOfSquares replied:
Wops this is just code I forgot to remove

applenick commented on core/src/main/java/tc/oc/pgm/stats/PlayerStatsInventoryMenuItem.java:
```diff
@@ -0,0 +1,135 @@
+public class PlayerStatsInventoryMenuItem implements InventoryMenuItem {
+
+  private final MatchPlayer player;
+  private final TextColor reset = TextColor.WHITE;
```
Make `RESET`

applenick commented on core/src/main/java/tc/oc/pgm/stats/TeamInventoryMenuItem.java:
```diff
@@ -0,0 +1,153 @@
+  private final Competitor team;
+  private final InventoryMenu teamSubGUI;
+
+  private final TextColor reset = TextColor.WHITE;
```
Make `RESET`

applenick reviewed (changes requested):
Hey King,
Just left some comments to help you improve some things. I look forward to the screenshots when you add them 😄 

Nice work!

applenick commented on core/src/main/java/tc/oc/pgm/menu/InventoryMenu.java:
```diff
@@ -1,50 +1,75 @@
   private final WeakHashMap<MatchPlayer, InventoryMenu> viewing =
       new WeakHashMap<>(); // Map of players who are viewing the gui, along with the menu
-  private final String title; // Title of the inventory
+  private final TranslatableComponent title; // Title of the inventory
```
```suggestion
  private final Component title; // Title of the inventory
```
Probably best to make this a Component, a menu title in the future may not come from a translation. Such as if the menu name was that of a username.

applenick commented on core/src/main/java/tc/oc/pgm/menu/InventoryMenu.java:
```diff
@@ -1,50 +1,75 @@
    */
-  public InventoryMenu(String title, int rows) {
+  public InventoryMenu(
+      Match match, TranslatableComponent title, int rows, List<InventoryMenuItem> items) {
```
See other comment

applenick reviewed (changes requested):
One more small change, but looks great besides that 👍 Nice job!

Pablete1234 commented on core/src/main/java/tc/oc/pgm/observers/ObserverToolsMatchModule.java:
```diff
@@ -52,11 +44,20 @@ public ObserverToolsMatchModule createMatchModule(Match match) throws ModuleLoad
     this.match = match;
-    this.menu = new ObserverToolMenu();
+
+    // TODO?: Add config options to enable/disable each tool
```
Unlikely, it would be easier to just deny via permissions them being used than to disable them/remove from menu

KingOfSquares replied:
This was leftover from @applenick , ill remove it

Pablete1234 commented on core/src/main/java/tc/oc/pgm/stats/PlayerStatsInventoryMenuItem.java:
```diff
@@ -0,0 +1,135 @@
+          TranslatableComponent.of(key, RESET, numberComponent(stat, TextColor.AQUA));
+      lore.add(TextTranslations.translateLegacy(loreComponent, bukkit));
+    }
+  }
```
Consider a method that takes `Number stat` as parameter, so both ints and doubles can be passed. Strings make no sense since it will literally throw an exception if you try a non-number string, so it makes no sense.
It would also eliminate the need for 2 "numberComponent" methods

Pablete1234 commented on core/src/main/java/tc/oc/pgm/stats/StatsMatchModule.java:
```diff
@@ -0,0 +1,492 @@
+    Instant longestFlagHoldCache;
+
+    //
+    private Future<?> task;
```
Give this a better name, what is the task for?

Pablete1234 commented on core/src/main/java/tc/oc/pgm/stats/StatsMatchModule.java:
```diff
@@ -0,0 +1,492 @@
+
+    final Collection<Competitor> competitors = match.getCompetitors();
+    // TODO: Support all Parties, Tributes(FFA)
+    if (verboseStats && competitors.stream().noneMatch(c -> c instanceof Tribute)) {
```
In reality, you just support teams not competitors, do a check on team match module existing or competitors being Team, instead of checking for no tributes

Pablete1234 commented on core/src/main/java/tc/oc/pgm/stats/StatsMatchModule.java:
```diff
@@ -0,0 +1,492 @@
+  private Component playerName(UUID playerUUID) {
+    return PlayerComponent.of(
+        Bukkit.getPlayer(playerUUID),
+        cachedUsernames.getOrDefault(playerUUID, "Unknown"),
```
There's a translatable for unknown, it's used in the tablist for unresolved authors, use it here

Pablete1234 commented on core/src/main/java/tc/oc/pgm/stats/StatsMatchModule.java:
```diff
@@ -430,17 +432,13 @@ public void onToolClick(PlayerInteractEvent event) {
+  public static Component numberComponent(Number stat, TextColor color) {
+    double doubleStat = stat.doubleValue();
+    if (doubleStat % 1 == 0) return TextComponent.of((int) stat, color, TextDecoration.BOLD);
+    else return TextComponent.of((double) stat, color, TextDecoration.BOLD);
```
Won't this create `3.333333333` things unless passed thru a number format?

KingOfSquares replied:
Its already passed through a format when its fetched from the `PlayerStats` 🤔  I'll check

KingOfSquares replied:
<img width="611" alt="Skjermbilde 2020-09-17 kl  18 49 54" src="https://user-images.githubusercontent.com/19822231/93502122-a02b2400-f916-11ea-970b-d8c10b4604d8.png">

<img width="516" alt="Skjermbilde 2020-09-17 kl  18 50 42" src="https://user-images.githubusercontent.com/19822231/93502182-b0db9a00-f916-11ea-9f1b-ed7a8759a728.png">

<img width="512" alt="Skjermbilde 2020-09-17 kl  18 50 37" src="https://user-images.githubusercontent.com/19822231/93502227-bc2ec580-f916-11ea-815c-435e4b1be771.png">

Had to do some small finicking to parse correctly, but should work as intended now 👍 
(Also fixed the fact damage to yourself is tracked in the last commit)

KingOfSquares commented:
Had to fix more stuff than I thought 😓 , hope its good now.

Pablete1234 commented on core/src/main/java/tc/oc/pgm/stats/TeamInventoryMenuItem.java:
```diff
@@ -0,0 +1,151 @@
+      shotsHit += stats.shotsHit;
+    }
+
+    double teamKD = teamDeaths == 0 ? teamKills : statFormat(teamKills / (double) teamDeaths);
+    // You hit 100% of the shots you dont take? :thinking:
+    double teamBowAcc = shotsTaken == 0 ? 100 : statFormat(shotsHit / (shotsTaken / (double) 100));
```
Consider leaving a `Double.NaN` in here, and on the component display as a `-` or similar, to show no number rather than 100% or 0%

Pablete1234 commented on core/src/main/java/tc/oc/pgm/stats/StatsMatchModule.java:
```diff
@@ -0,0 +1,489 @@
+    /** A common format used by all stats with decimals */
+    public static final DecimalFormat STATS_DECIMALFORMAT = new DecimalFormat("#.##");
+
+    public static double statFormat(Number number) {
+      String stat = STATS_DECIMALFORMAT.format(number).replace(',', '.');
+      return Double.parseDouble(stat);
+    }
```
Don't convert a double into a double. Format when you're about to display it to the user, when creating the TextComponent. Here's you're not setting format, you're just arbitrarily altering doubles to try to force a specific rounding

Pablete1234 commented on core/src/main/java/tc/oc/pgm/stats/StatsMatchModule.java:
```diff
@@ -0,0 +1,489 @@
+  public static Component numberComponent(Number stat, TextColor color) {
+    double doubleStat = stat.doubleValue();
+    if (doubleStat % 1 == 0) return TextComponent.of(stat.intValue(), color, TextDecoration.BOLD);
+    else return TextComponent.of(doubleStat, color, TextDecoration.BOLD);
```
Use a number formatter here, to set the format you want (with 2 decimals) and pass the resulting formatted string to TextComponent

applenick commented on util/src/main/java/tc/oc/pgm/util/text/types/PlayerComponent.java:
```diff
@@ -92,7 +92,8 @@ static Component of(
   // What an offline or vanished username renders as
   static TextComponent.Builder formatOffline(String name, boolean plain) {
-    TextComponent.Builder component = TextComponent.builder().append(name);
+    TextComponent.Builder component =
+        TextComponent.builder().append(TranslatableComponent.of(name));
```
What's the reason behind wrapping the string in a translatable component? 🤔

KingOfSquares replied:
Right, I wanted to use a translatable string key, but I realize I should require a `Component` instead of a `String` there in the first place. (Instead of hardcoded "Unknown" as it was previously I wanted "misc.unknown")
I will revert this change and do a separate PR for this specific issue

KingOfSquares commented:
Thanks for feedback! I decided to move out `PlayerStats` to its own class now that `StatsMatchModule` has more code + it has its own package now anyways. With that change I also changed how the stat values are accessed, now through proper getters.

Pablete1234 commented on core/src/main/java/tc/oc/pgm/stats/PlayerStats.java:
```diff
@@ -0,0 +1,188 @@
+  }
+
+  public double getArrowAccuracy() {
+    double acc;
+    if (shotsTaken == 0) acc = Double.NaN;
+    else acc = shotsHit / (shotsTaken / (double) 100);
+    return acc;
```
```suggestion
    if (shotsTaken == 0) return Double.NaN;
    return shotsHit / (shotsTaken / (double) 100);
```
Simplify logic

Pablete1234 commented on core/src/main/java/tc/oc/pgm/stats/PlayerStats.java:
```diff
@@ -0,0 +1,188 @@
+  // Getters, both raw stats and some handy calculations
+
+  public double getKD() {
+    double kd;
+    if (deaths == 0) {
+      kd = kills;
+    } else {
+      kd = kills / (double) deaths;
+    }
+    return kd;
```
```suggestion
return kills / Math.max(1d, (double) deaths);
```
Simplify this, you may be able to get rid of the double cast too if the first parameter is explicitly double

KingOfSquares replied:
It will(is what my IDE says)! Since `Math.max` returns a double in that case

Pablete1234 commented on core/src/main/java/tc/oc/pgm/stats/StatsMatchModule.java:
```diff
@@ -0,0 +1,377 @@
+  public static Component numberComponent(Number stat, TextColor color) {
+    double doubleStat = stat.doubleValue();
+    String returnValue;
+    if (Double.isNaN(stat.doubleValue()))
```
```suggestion
    if (Double.isNaN(doubleStat))
```

Pablete1234 reviewed (commented):
Just a couple simplifications, otherwise lgtm

KingOfSquares commented:
Thanks @Pablete1234 ❤️

Pugzy commented on core/src/main/java/tc/oc/pgm/stats/TeamStatsInventoryMenuItem.java:
```diff
@@ -0,0 +1,149 @@
+
+  @Override
+  public Material getMaterial(MatchPlayer player) {
+    return Material.WOOL;
```
Coloured leather helm to match the team picker pls 🙏

Pablete1234 replied:
I would not make this a helmet, people will think it's the team picker. Maybe chestplate or leggings (boots should be used for leaving even tho that's not on this version of pgm)

KingOfSquares replied:
I think leggings look weird, but chest plates is a good idea!

Electroid commented on util/src/main/i18n/templates/match.properties:
```diff
@@ -52,19 +52,56 @@ match.stats = Kills: {0}  |  Killstreak: {1}  |  Deaths: {2}  |  K/D: {3}
 match.stats.killstreak = Killstreak: {1} by {0}
 
+# {0} = a killstreak(number)
+match.stats.killstreak.concise = Highest killstreak: {0}
```
`Best Killstreak`

Electroid commented on util/src/main/i18n/templates/match.properties:
```diff
@@ -52,19 +52,56 @@ match.stats = Kills: {0}  |  Killstreak: {1}  |  Deaths: {2}  |  K/D: {3}
+# {0} = the number of damage done
+# {1} = the number of damage done with a bow
+# {2} = the number of damage taken
+match.stats.damage.concise = Damage given(bow) / taken: {0}({1}) / {2}
```
At face value this looks like a very confusing stat, not even sure what it means. Is it nessecary? Seems Bow accuracy is fine enough?

KingOfSquares replied:
This verbose stat implementation was made with more competitive players in mind, and this is something that can be useful to analyze performance in a comp match. This is also why I have decided to not pay attention to the stat setting while verbose stats is enabled

Damage given: All damage given by the player
(bow): All damage done by bow and arrow  by this player
taken: All damage taken by this player

I have yet to see how big numbers realistically get during a full match, since I only get to test on myself. But it is something that I hope to change if necessary after some testing

Electroid replied:
For the first version of verbose stats, let's leave this out. I'm ok considering it later.

Electroid replied:
Commit your changes and I'll review again

Commit:
Verbose stats
- Rip commit history

- Also includes a refactor of InventoryMenu

Signed-off-by: KingSimon <simonmorland@gmail.com>

Fix some logic and prepare for longestFlagHold

Signed-off-by: KingSimon <simonmorland@gmail.com>

Add longest flag hold
- some pretty stuff

Signed-off-by: KingSimon <simonmorland@gmail.com>

Even prettier

Signed-off-by: KingSimon <simonmorland@gmail.com>

Ready for release

Signed-off-by: KingSimon <simonmorland@gmail.com>

Remove old code

Signed-off-by: KingSimon <simonmorland@gmail.com>

Reviews and some javadoc

Signed-off-by: KingSimon <simonmorland@gmail.com>

Make inventory title a Component
 instead of TranslatableComponent

Signed-off-by: KingSimon <simonmorland@gmail.com>

Verbose stats
- Rip commit history

- Also includes a refactor of InventoryMenu

Signed-off-by: KingSimon <simonmorland@gmail.com>

Fix some logic and prepare for longestFlagHold

Signed-off-by: KingSimon <simonmorland@gmail.com>

Add longest flag hold
- some pretty stuff

Signed-off-by: KingSimon <simonmorland@gmail.com>

Even prettier

Signed-off-by: KingSimon <simonmorland@gmail.com>

Ready for release

Signed-off-by: KingSimon <simonmorland@gmail.com>

Remove old code

Signed-off-by: KingSimon <simonmorland@gmail.com>

Reviews and some javadoc

Signed-off-by: KingSimon <simonmorland@gmail.com>

Cleanup
- TranslatableComponent -> Component
- Remove some old debug stuff
- Fix some typos and javadoc stuff

Signed-off-by: KingSimon <simonmorland@gmail.com>

Pass all formatted doubles as doubles instead of Strings
- Determine if a number is an int or double in StatsMatchModule#numberComponent

Signed-off-by: KingSimon <simonmorland@gmail.com>

Use translatable instead of String
- Rename hotbar task in PlayerStats

Signed-off-by: KingSimon <simonmorland@gmail.com>

Only display verbose stats if all competitors are Teams
-Format

Signed-off-by: KingSimon <simonmorland@gmail.com>

Correctly parse doubles++
- Fix page bug in InventoryMenuUtils#prettyMenu
- Only display damage done if verbose stats is enabled
- fix bad code in StatsMatchModule#getBasicStatsMessage
- fix bad casting in StatsMatchModule#numberComponent
- make sure damages used in StatsMatchModule#onMatchEnd uses the formatted damage numbers
- limit ability to fetch unformatted doubles from PlayerStats
- Fix flag stats

Signed-off-by: KingSimon <simonmorland@gmail.com>

Move formatting of doubles
- Into StatsMatchModule#numberComponent
- If a stat is NaN it is displayed as "-"
- Revert change to PlayerComponent(will be its own PR)

Signed-off-by: KingSimon <simonmorland@gmail.com>

Move out PlayerStats to its own class
- change access to stat values

Signed-off-by: KingSimon <simonmorland@gmail.com>

TeamInventoryMenuItem -> TeamStatsInventoryMenuItem

Signed-off-by: KingSimon <simonmorland@gmail.com>

Simplify some logic

Signed-off-by: KingSimon <simonmorland@gmail.com>

Change the icon for team stats to colored chestplates

Signed-off-by: KingSimon <simonmorland@gmail.com>

Clean up

Signed-off-by: KingSimon <manmusic979@gmail.com>
Signed-off-by: KingSimon <simonmorland@gmail.com>

Electroid commented on util/src/main/i18n/templates/misc.properties:
```diff
@@ -114,6 +114,10 @@ misc.to = To
 
 misc.from = From
 
+misc.nextPage = Next page
```
Next Page

Electroid commented on util/src/main/i18n/templates/misc.properties:
```diff
@@ -114,6 +114,10 @@ misc.to = To
 
+misc.nextPage = Next page
+
+misc.previousPage = Previous page
```
Previous Page

Electroid commented on util/src/main/i18n/templates/match.properties:
```diff
@@ -52,19 +52,56 @@ match.stats = Kills: {0}  |  Killstreak: {1}  |  Deaths: {2}  |  K/D: {3}
+match.stats.broken.concise = Monument pieces broken: {0}
+
+# {0} = a player name
+match.stats.team = {0}'s Match Stats
```
I think this should just be `{0}'s Stats`

Electroid commented on util/src/main/i18n/templates/match.properties:
```diff
@@ -52,19 +52,56 @@ match.stats = Kills: {0}  |  Killstreak: {1}  |  Deaths: {2}  |  K/D: {3}
+match.stats.bow.concise = Bow shots / Accuracy: {0} / {1}%
+
+# {0} = the number of flags captured
+match.stats.flagsCaptured.concise = Flags captured: {0}
```
Flags Captured

Electroid commented on util/src/main/i18n/templates/match.properties:
```diff
@@ -52,19 +52,56 @@ match.stats = Kills: {0}  |  Killstreak: {1}  |  Deaths: {2}  |  K/D: {3}
+match.stats.flagsCaptured.concise = Flags captured: {0}
+
+# {0} = a number representing the duration of the hold
+match.stats.flaghold.concise = Longest flag hold: {0}
```
Longest Flag Capture

Electroid commented on util/src/main/i18n/templates/match.properties:
```diff
@@ -52,19 +52,56 @@ match.stats = Kills: {0}  |  Killstreak: {1}  |  Deaths: {2}  |  K/D: {3}
+match.stats.flaghold.concise = Longest flag hold: {0}
+
+# {0} the amount of monument pieces broken(number)
+match.stats.broken.concise = Monument pieces broken: {0}
```
Monument Blocks Broken

Electroid commented on util/src/main/i18n/templates/match.properties:
```diff
@@ -52,19 +52,56 @@ match.stats = Kills: {0}  |  Killstreak: {1}  |  Deaths: {2}  |  K/D: {3}
-match.stats.overall = Match Stats
+# {0} = a player name
+# {1} = an amount of damage
+match.stats.damage = Damage: {1} by {0}
```
Melee Damage

Electroid replied:
Actually, not sure what this is for?

KingOfSquares replied:
This would appear on the "Best of" stats at the end of the match.
It takes all damage done by that player towards other players.

Electroid commented on util/src/main/i18n/templates/match.properties:
```diff
@@ -52,19 +52,56 @@ match.stats = Kills: {0}  |  Killstreak: {1}  |  Deaths: {2}  |  K/D: {3}
+
+# {0} = the number of bow shots taken
+# {1} = the number representing someones accuracy(e.g 35%)
+match.stats.bow.concise = Bow shots / Accuracy: {0} / {1}%
```
Bow Shots / Accuracy

Electroid reviewed (changes requested):
The code looks good, the translations need some tweaking.

KingOfSquares committed:
Fix destroyable stat
- remove detour in InventoryMenu

Signed-off-by: KingSimon <19822231+KingOfSquares@users.noreply.github.com>

KingOfSquares committed:
Fix translations

Signed-off-by: KingSimon <19822231+KingOfSquares@users.noreply.github.com>

KingOfSquares committed:
Revert "Remove verbose damage stat"

This reverts commit 7ad26b070c44e7062a32d2d06ab13806edec45c8.

KingOfSquares committed:
Remove verbose damage stat

Signed-off-by: KingSimon <19822231+KingOfSquares@users.noreply.github.com>

KingOfSquares committed:
Shorten big numbers
                                                         - Stats like damage can get really big and needs to be shortened to fit on the screen
                                                         - Tweak the translation syntax of the verbose damage stat

Signed-off-by: KingSimon <19822231+KingOfSquares@users.noreply.github.com>

KingOfSquares commented:
No idea why the DCO is complaining..

KingOfSquares committed:
Actually fix translation

Signed-off-by: KingSimon <19822231+KingOfSquares@users.noreply.github.com>
